### ISSUE TITLE
sha512 support?

### ISSUE THREAD

**Issue by 0x5a17ed** [minio/sha256-simd#72]
Hi,

This package looks very promising, but it seems to me that support for the `sha512` algorithm is missing. Unfortunately, if it wasn't for my lack of expertise in implementing any algorithm of the `sha` family, I would have tried tackling this myself. Is there any chance that `sha512` can be accelerated with the AVX instruction set as well?

**Comment by harshavardhana** [comment]
this is a sha256 package, why would this be doing sha512.

**Comment by 0x5a17ed** [comment]
Again, I'm not an expert, but as far as I understand the matter, both `sha256` and `sha512` are based on the same algorithm classified as the `sha-2` family of hashing algorithms and their only difference is the word size in the hashing step and their output length. 

So, technically speaking, they are basically the same algorithm and therefore my train of thought was that a package implementing one of them might be a good spot to provide an implementation for the other one as well. The name of this package in particular, of course, only indicates that the `sha256` variant is implemented.

**Comment by AudriusButkevicius** [comment]
I might be wrong, but I think this package is effectively obsolete as the standard library implementation already picked up these speedups.

Also, the fact that you think it's easy doesn't suddenly make someone want to spend their time doing it.

**Comment by klauspost** [comment]
Correct, it is obsolete.

**Comment by 0x5a17ed** [comment]
Thanks everyone for the input and I apologize for the disturbance caused by this issue. 

I'm not sure what I was looking at before when I was going through the implementation in the standard library, but given the hint that the implementation is now part of the standard library I found the correct implementation here: <https://github.com/golang/go/blob/956f8a67dd7319bad60c015d982f0b2e95b9f382/src/crypto/sha512/sha512block_amd64.s#L277>.

> Also, the fact that you think it's easy doesn't suddenly make someone want to spend their time doing it.

English is not my native language but I never intended or tried to say it would be easy to extend the implementation in this package to provide an `sha512` function. I tried to admit that myself in the first comment at the beginning of the issue that it's hard, especially in assembly, and I'm nowhere close to being able to implement the algorithm myself. The point I tried to make was that on paper the difference between sha256 and sha512 is their word size and therefore are close enough in proximity to justify both being implemented in the same package.
